recipes-extended/xen/files/xsm-for-vtd.patch

```diff
@@ -0,0 +1,259 @@
+SHORT DESCRIPTION:
+################################################################################
+Enable XSM control over device passthrough with separate policy controls for:
+* with and without an IOMMU available
+* an IOMMU with more capable isolation, in this case via interrupt remapping
+
+################################################################################
+LONG DESCRIPTION:
+################################################################################
+
+################################################################################
+CHANGELOG
+################################################################################
+
+################################################################################
+REMOVAL
+################################################################################
+Via upstreaming.
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+Yes.
+
+################################################################################
+INTERNAL DEPENDENCIES
+################################################################################
+OpenXT's XSM policy in xsm-policy.git
+
+################################################################################
+PATCHES
+################################################################################
+diff --git a/xen/xsm/flask/hooks.c b/xen/xsm/flask/hooks.c
+index 59905e2..bac29ad 100644
+--- a/xen/xsm/flask/hooks.c
++++ b/xen/xsm/flask/hooks.c
+@@ -20,6 +20,7 @@
+ #include <xen/errno.h>
+ #include <xen/guest_access.h>
+ #include <xen/xenoprof.h>
++#include <xen/iommu.h>
+ #ifdef HAS_PCI
+ #include <asm/msi.h>
+ #endif
+@@ -866,11 +867,41 @@ static int flask_map_domain_msi (struct domain *d, int irq, void *data,
+ #endif
+ }
+ 
++static u32 flask_iommu_resource_use_perm(void)
++{
++    /* This function defines the permission level required for allowing
++     * domains to use assigned devices.
++     *
++     * Interrupt remapping is an essential property for ensuring strict
++     * isolation of devices.
++     *
++     * If strict isolation is not available then check the platform permission
++     * to see whether the policy allows use of a less capable IOMMU.
++     */
++    struct av_decision avd;
++    u32 csid = domain_sid(current->domain);
++
++    /* Use the noaudit method for checking this permission because
++     * we're identifying which _other_ permission to check for.
++     * Failure here is sometimes expected and always acceptable so
++     * it is important not to log it to protect the logs.
++     */
++    if (iommu_enabled && (iommu_intremap ||
++                          !avc_has_perm_noaudit(csid, SECINITSID_XEN,
++                                                SECCLASS_IOMMU,
++                                                IOMMU__USE_INSECURE, &avd)))
```

Comment:
Stephen, Daniel, thanks very much for taking the time to review this and for your comments.

### Re: per-domain permissions:

> Does the iommu use_insecure permission check make sense as a per-domain permission?
> Would we ever allow it for domain D1 but deny it for domain D2?

It's a good question and I think possibly, or possibly not.
Here is a somewhat contrived example but perhaps:

D1: contains the graphics display device driver and a local display device is assigned.
Exposed to internal graphical output only.

IOMMU containment ensures that a buggy device driver is less able to impact system stability, or usability, or the integrity of other VMs.

D2: contains a NIC device driver and an external network interface is assigned.
Exposed to external network traffic.

IOMMU containment ensures that a compromised device driver domain cannot be used by an attacker to obtain access to other data on the system.

(ie. the less capable IOMMU is deemed acceptable for D1 but not D2)

The same host software (and same XSM policy) can be used on multiple machines with different IOMMU capabilities, with the behaviour on the hosts with weaker IOMMUs being that VMs are operable with working display output, but they are isolated from the external network by disallowing the NIC device assignment.

### Re: design change to use three distinct use_with_* bits :
I will do that. I'll prepare an alternative PR with that method for comparison with this one.

With this version of the code, three points that I like -- and will hopefully keep in the other version -- are:

1) The "allow use of the insecure IOMMU" capability is confined to only the domains expressly indicated in the policy, and all others denied, which removes a capability from domains that have no need of it.

2) The XSM policy alone, rather than a combination of XSM policy and the host command line, can set the isolation requirements.

Setting the IOMMU capability requirement via the boot command line makes the configuration setting more distant from the VM XSM policy which depends upon it, in a different format for a different tool.

In OpenXT, we want to support:

* Producing a single build of software that runs on diverse hardware -- so default to not requiring specific hardware capabilities.

* Allow derivative projects to produce specialized builds, to _require_ that specific hardware-enabled security features are always present and enabled, and accept a corresponding reduction in the hardware compatibility.

If separate builds are being produced to use with different hardware, swapping a XSM policy seems simpler to manage than rewriting the hypervisor command lines in the bootloader config files and then ensuring that changed settings are preserved correctly across software upgrades. It's still a change to a boot-time file though, so maybe it's not all that different in practice.

3) The XSM policy and iommu control variables are tested every time the device assignment action is performed, and not only relying on a one-time variable check at host boot.

Hope that explains some of the design, and I'll get a revised version prepared next. Thanks!